### ISSUE TITLE
NGA FY22: refactor OpenMP backend to use more native OpenMP constructs

### ISSUE THREAD

**Issue by fnrizzi** [kokkos/kokkos#4369]
Some of the items to refactor/address: 
- [x] `parallel_for(RangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
- [x] `parallel_for(MDRangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
- [ ] `parallel_reduce(RangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static) 
  (follow OpenMPTarget backend for custom reductions)
- [ ] `parallel_reduce(MDRangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static) 
  (follow OpenMPTarget backend for custom reductions)

**Comment by fnrizzi** [comment]
More comments on this coming soon.

**Comment by cz4rs** [comment]
> Some of the items to refactor/address:
> 
>     * [ ]  `parallel_for(RangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
> 
>     * [ ]  `parallel_for(MDRangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
> 
>     * [ ]  `parallel_reduce(RangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
>       (follow OpenMPTarget backend for custom reductions)
> 
>     * [ ]  `parallel_reduce(MDRangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
>       (follow OpenMPTarget backend for custom reductions)

Current status:
- [x] `parallel_for(RangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
  - native implementation: https://github.com/kokkos/kokkos/pull/4664
- [x] `parallel_for(MDRangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
  - native implementation: https://github.com/kokkos/kokkos/pull/4664
- [ ] `parallel_reduce(RangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static) 
  - Work in progress following OpenMPTarget backend implementation.
- [ ] `parallel_reduce(MDRangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static) 
  - No work for this one yet, expected to be very similar to `parallel_reduce(RangePolicy)`

Questions:
- Are there any known issues with the OpenMPTarget implementation for `parallel_reduce`?
- Was there any feedback for https://github.com/kokkos/kokkos/pull/4119?
> need to hear back from Uintah before we merge

**Comment by cz4rs** [comment]
Refactored version of `parallel_for` for both `RangePolicy` and `MDRangePolicy` was implemented with #4664.

**Comment by dalg24** [comment]
> Refactored version of `parallel_for` for both `RangePolicy` and `MDRangePolicy` was implemented with #4664.

The description includes `parallel_reduce`.

**Comment by fnrizzi** [comment]
yes i think seeing from the history above was picked up by Phil and then removed , let's reopen it

**Comment by fnrizzi** [comment]
@cz4rs can you look a bit into this?

**Comment by cz4rs** [comment]
> @cz4rs can you look a bit into this?

Sure, I have some leftovers from the work on `parallel_reduce` on my branch too. Let me get back into it and see what's missing.
